wire-ios-sync-engine/Source/Synchronization/Strategies/UserClientRequestStrategy.swift

```diff
@@ -158,6 +160,23 @@ public final class UserClientRequestStrategy: ZMObjectSyncStrategy, ZMObjectStra
             }
         }
 
+        if clientRegistrationStatus.currentPhase == .generatingPrekeys {
+            return nil
+        }
+
+        if clientRegistrationStatus.currentPhase == .waitingForPrekeys {
+            clientRegistrationStatus.willGeneratePrekeys()
+            let groups = managedObjectContext?.enterAllGroupsExceptSecondary()
+            Task {
+                let prekeys = try await prekeyGenerator.generatePrekeys()
+                let lastResortPrekey = try await prekeyGenerator.generateLastResortPrekey()
+                await managedObjectContext?.perform {
+                    clientRegistrationStatus.didGeneratePrekeys(prekeys, lastResortPrekey: lastResortPrekey)
+                }
+                managedObjectContext?.leaveAllGroups(groups)
+            }
```

Comment:
```suggestion
            if let managedObjectContext {
            WaitingGroupTask(context: managedObjectContext) {
                let prekeys = try await prekeyGenerator.generatePrekeys()
                let lastResortPrekey = try await prekeyGenerator.generateLastResortPrekey()
                await managedObjectContext?.perform {
                    clientRegistrationStatus.didGeneratePrekeys(prekeys, lastResortPrekey: lastResortPrekey)
                }
            }
            }
```

Reply:
@typfel 

1) if it throws the clientRegistrationStatus is not called -> do we need a `clientRegistrationStatus.didFailGenerate...` ?
and report to the user?

+ 2) the test don't exit the group right?

Reply:
Good point, we should definitely do something here. But what can do we except for logging the error? If it fails to generate prekey something is very wrong and I don't except it work by just re-trying.

Reply:
ok added logging for now, I'll keep this in mind